File: python/api_requests.py
"""
Python API Request Performance Test
Uses asyncio and aiohttp for efficient concurrent HTTP requests.
Collects comprehensive metrics including response times, throughput, and percentiles.
"""

import asyncio
import aiohttp
import time
import json
import statistics
from pathlib import Path

async def make_request(session, url, request_id, start_time):
    """
    Make a single HTTP request asynchronously.
    Returns timing data for the request.
    """
    request_start = time.time()
    try:
        async with session.get(url, timeout=aiohttp.ClientTimeout(total=10)) as response:
            await response.text()  # Read response body
            request_end = time.time()
            return {
                'id': request_id,
                'success': response.status == 200,
                'response_time_ms': (request_end - request_start) * 1000,
                'timestamp': request_end - start_time
            }
    except Exception as e:
        request_end = time.time()
        return {
            'id': request_id,
            'success': False,
            'response_time_ms': (request_end - request_start) * 1000,
            'timestamp': request_end - start_time,
            'error': str(e)
        }

async def run_benchmark():
    """
    Run API request benchmark with 1,000 concurrent requests.
    Uses aiohttp for async HTTP requests with connection pooling.
    """
    url = "https://jsonplaceholder.typicode.com/posts/1"
    num_requests = 1000
    
    print(f"Starting benchmark: {num_requests} requests to {url}")
    
    # Measure total time for concurrent requests
    start_time = time.time()
    
    # Create a connection pool with appropriate limits
    connector = aiohttp.TCPConnector(limit=50, limit_per_host=50)
    async with aiohttp.ClientSession(connector=connector) as session:
        tasks = [make_request(session, url, i, start_time) for i in range(num_requests)]
        results = await asyncio.gather(*tasks)
    
    end_time = time.time()
    total_time = end_time - start_time
    
    # Analyze results
    successful_results = [r for r in results if r['success']]
    failed_results = [r for r in results if not r['success']]
    
    successful_count = len(successful_results)
    failed_count = len(failed_results)
    
    # Extract response times from successful requests
    response_times = [r['response_time_ms'] for r in successful_results]
    
    # Calculate metrics
    if response_times:
        sorted_times = sorted(response_times)
        metrics = {
            'language': 'python',
            'total_requests': num_requests,
            'successful_requests': successful_count,
            'failed_requests': failed_count,
            'total_time_seconds': round(total_time, 2),
            'requests_per_second': round(num_requests / total_time, 2),
            'response_times': {
                'min_ms': round(min(response_times), 2),
                'max_ms': round(max(response_times), 2),
                'average_ms': round(statistics.mean(response_times), 2),
                'median_ms': round(statistics.median(response_times), 2),
                'p95_ms': round(sorted_times[int(len(sorted_times) * 0.95)], 2),
                'p99_ms': round(sorted_times[int(len(sorted_times) * 0.99)], 2)
            },
            'timeseries': [
                {
                    'timestamp': round(r['timestamp'], 3),
                    'response_time_ms': round(r['response_time_ms'], 2)
                }
                for r in successful_results
            ]
        }
    else:
        metrics = {
            'language': 'python',
            'total_requests': num_requests,
            'successful_requests': 0,
            'failed_requests': failed_count,
            'total_time_seconds': round(total_time, 2),
            'requests_per_second': 0,
            'response_times': {},
            'timeseries': []
        }
    
    # Save results to JSON file
    output_file = Path(__file__).parent / 'api_results.json'
    with open(output_file, 'w') as f:
        json.dump(metrics, f, indent=2)
    
    print(f"\nResults saved to {output_file}")
    
    return metrics

def print_results(metrics):
    """Print formatted results to console."""
    print(f"\n{'='*60}")
    print(f"Language: {metrics['language'].upper()}")
    print(f"{'='*60}")
    print(f"Total Requests: {metrics['total_requests']}")
    print(f"Successful: {metrics['successful_requests']}")
    print(f"Failed: {metrics['failed_requests']}")
    print(f"Total Time: {metrics['total_time_seconds']:.2f}s")
    print(f"Requests/sec: {metrics['requests_per_second']:.2f}")
    
    if metrics['response_times']:
        print(f"\nResponse Times (ms):")
        print(f"  Min: {metrics['response_times']['min_ms']:.2f}")
        print(f"  Max: {metrics['response_times']['max_ms']:.2f}")
        print(f"  Avg: {metrics['response_times']['average_ms']:.2f}")
        print(f"  Median: {metrics['response_times']['median_ms']:.2f}")
        print(f"  P95: {metrics['response_times']['p95_ms']:.2f}")
        print(f"  P99: {metrics['response_times']['p99_ms']:.2f}")
    print(f"{'='*60}\n")

if __name__ == '__main__':
    metrics = asyncio.run(run_benchmark())
    print_results(metrics)

File: javascript/api_requests.js
/**
 * JavaScript/Node.js API Request Performance Test
 * Uses native https module with Promise.all for concurrent requests.
 * Collects comprehensive metrics including response times, throughput, and percentiles.
 */

const https = require('https');
const fs = require('fs');
const path = require('path');

/**
 * Make a single HTTP request and track timing.
 */
function makeRequest(url, requestId, startTime) {
    const requestStart = performance.now();
    
    return new Promise((resolve) => {
        https.get(url, (res) => {
            let data = '';
            res.on('data', (chunk) => data += chunk);
            res.on('end', () => {
                const requestEnd = performance.now();
                resolve({
                    id: requestId,
                    success: res.statusCode === 200,
                    response_time_ms: requestEnd - requestStart,
                    timestamp: (requestEnd - startTime) / 1000
                });
            });
        }).on('error', (err) => {
            const requestEnd = performance.now();
            resolve({
                id: requestId,
                success: false,
                response_time_ms: requestEnd - requestStart,
                timestamp: (requestEnd - startTime) / 1000,
                error: err.message
            });
        });
    });
}

/**
 * Calculate percentile from sorted array.
 */
function percentile(arr, p) {
    if (arr.length === 0) return 0;
    const index = Math.ceil(arr.length * p) - 1;
    return arr[Math.max(0, Math.min(index, arr.length - 1))];
}

/**
 * Run API request benchmark with 1,000 concurrent requests.
 * Uses native https module with connection pooling.
 */
async function runBenchmark() {
    const url = "https://jsonplaceholder.typicode.com/posts/1";
    const numRequests = 1000;
    
    console.log(`Starting benchmark: ${numRequests} requests to ${url}`);
    
    // Increase max sockets for better concurrency
    https.globalAgent.maxSockets = 50;
    
    // Measure total time for concurrent requests
    const startTime = performance.now();
    
    const promises = Array.from({ length: numRequests }, (_, i) => 
        makeRequest(url, i, startTime)
    );
    
    const results = await Promise.all(promises);
    
    const endTime = performance.now();
    const totalTime = (endTime - startTime) / 1000; // Convert to seconds
    
    // Analyze results
    const successfulResults = results.filter(r => r.success);
    const failedResults = results.filter(r => !r.success);
    
    const successfulCount = successfulResults.length;
    const failedCount = failedResults.length;
    
    // Extract response times from successful requests
    const responseTimes = successfulResults.map(r => r.response_time_ms);
    
    let metrics;
    
    if (responseTimes.length > 0) {
        const sortedTimes = responseTimes.slice().sort((a, b) => a - b);
        const sum = responseTimes.reduce((acc, val) => acc + val, 0);
        
        metrics = {
            language: 'javascript',
            total_requests: numRequests,
            successful_requests: successfulCount,
            failed_requests: failedCount,
            total_time_seconds: parseFloat(totalTime.toFixed(2)),
            requests_per_second: parseFloat((numRequests / totalTime).toFixed(2)),
            response_times: {
                min_ms: parseFloat(Math.min(...responseTimes).toFixed(2)),
                max_ms: parseFloat(Math.max(...responseTimes).toFixed(2)),
                average_ms: parseFloat((sum / responseTimes.length).toFixed(2)),
                median_ms: parseFloat(percentile(sortedTimes, 0.5).toFixed(2)),
                p95_ms: parseFloat(percentile(sortedTimes, 0.95).toFixed(2)),
                p99_ms: parseFloat(percentile(sortedTimes, 0.99).toFixed(2))
            },
            timeseries: successfulResults.map(r => ({
                timestamp: parseFloat(r.timestamp.toFixed(3)),
                response_time_ms: parseFloat(r.response_time_ms.toFixed(2))
            }))
        };
    } else {
        metrics = {
            language: 'javascript',
            total_requests: numRequests,
            successful_requests: 0,
            failed_requests: failedCount,
            total_time_seconds: parseFloat(totalTime.toFixed(2)),
            requests_per_second: 0,
            response_times: {},
            timeseries: []
        };
    }
    
    // Save results to JSON file
    const outputFile = path.join(__dirname, 'api_results.json');
    fs.writeFileSync(outputFile, JSON.stringify(metrics, null, 2));
    
    console.log(`\nResults saved to ${outputFile}`);
    
    return metrics;
}

/**
 * Print formatted results to console.
 */
function printResults(metrics) {
    console.log('\n' + '='.repeat(60));
    console.log(`Language: ${metrics.language.toUpperCase()}`);
    console.log('='.repeat(60));
    console.log(`Total Requests: ${metrics.total_requests}`);
    console.log(`Successful: ${metrics.successful_requests}`);
    console.log(`Failed: ${metrics.failed_requests}`);
    console.log(`Total Time: ${metrics.total_time_seconds.toFixed(2)}s`);
    console.log(`Requests/sec: ${metrics.requests_per_second.toFixed(2)}`);
    
    if (Object.keys(metrics.response_times).length > 0) {
        console.log('\nResponse Times (ms):');
        console.log(`  Min: ${metrics.response_times.min_ms.toFixed(2)}`);
        console.log(`  Max: ${metrics.response_times.max_ms.toFixed(2)}`);
        console.log(`  Avg: ${metrics.response_times.average_ms.toFixed(2)}`);
        console.log(`  Median: ${metrics.response_times.median_ms.toFixed(2)}`);
        console.log(`  P95: ${metrics.response_times.p95_ms.toFixed(2)}`);
        console.log(`  P99: ${metrics.response_times.p99_ms.toFixed(2)}`);
    }
    console.log('='.repeat(60) + '\n');
}

if (require.main === module) {
    runBenchmark().then(metrics => {
        printResults(metrics);
    }).catch(err => {
        console.error('Benchmark failed:', err);
        process.exit(1);
    });
}

module.exports = { runBenchmark };
